lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp

```diff
@@ -463,93 +463,78 @@ LogicalResult static applyNoBlackBoxStyleDataTaps(const AnnoPathValue &target,
-        refSendBuilder.setInsertionPointAfter(srcTarget->ref.getOp());
+        sendVal = srcTarget->ref.getImpl().getOp()->getResult(0);
+        sendBuilder.setInsertionPointAfter(srcTarget->ref.getOp());
       } else if (srcTarget->ref.getImpl().isPort()) {
-        refSendBase =
-            srcModule.getArgument(srcTarget->ref.getImpl().getPortNo());
-        refSendBuilder.setInsertionPointToStart(srcModule.getBodyBlock());
+        sendVal = srcModule.getArgument(srcTarget->ref.getImpl().getPortNo());
+        sendBuilder.setInsertionPointToStart(srcModule.getBodyBlock());
       }
       // If the target value is a field of an aggregate create the
       // subfield/subaccess into it.
-      refSendBase =
-          getValueByFieldID(refSendBuilder, refSendBase, srcTarget->fieldIdx);
-      // Note: No DontTouch added to refSendTarget, it can be constantprop'ed or
+      sendVal = getValueByFieldID(sendBuilder, sendVal, srcTarget->fieldIdx);
+      // Note: No DontTouch added to sendVal, it can be constantprop'ed or
       // CSE'ed.
-      sendVal = refSendBuilder.create<RefSendOp>(refSendBase);
     }
-    // Now drill ports to connect the `sendVal` to the `wireTarget`.
-    auto remoteXMR = borePortsOnPath(
-        pathFromSrcToWire, lcaModule, sendVal, tapName.getValue(),
-        state.instancePathCache,
-        [&](FModuleLike mod) -> ModuleNamespace & {
-          return state.getNamespace(mod);
-        },
-        &state.targetCaches);
-    ImplicitLocOpBuilder refResolveBuilder(wireModule.getLoc(), wireModule);
+
+    // Mark sink module "NoDedup".
+    // Unique paths are needed for RefType, deduplication may break that.
     AnnotationSet annos(wireModule);
     if (!annos.hasAnnotation(noDedupAnnoClassName)) {
       annos.addAnnotations(noDedupAnno);
       annos.applyToOperation(wireModule);
     }
 
-    if (remoteXMR.isa<BlockArgument>())
-      refResolveBuilder.setInsertionPointToStart(wireModule.getBodyBlock());
-    else
-      refResolveBuilder.setInsertionPointAfter(remoteXMR.getDefiningOp());
-    auto refResolve = refResolveBuilder.create<RefResolveOp>(remoteXMR);
-    refResolveBuilder.setInsertionPointToEnd(
-        wireTarget->ref.getOp()->getBlock());
-    auto wireType = wireTarget->ref.getOp()
-                        ->getResult(0)
-                        .getType()
-                        .cast<FIRRTLType>()
-                        .cast<FIRRTLBaseType>();
-    Value resolveResult = refResolve.getResult();
-    if (refResolve.getType().isResetType() &&
-        refResolve.getType().getWidthlessType() !=
-            wireType.getWidthlessType()) {
-      if (wireType.dyn_cast<IntType>())
-        resolveResult =
-            refResolveBuilder.create<AsUIntPrimOp>(wireType, refResolve);
-      else if (wireType.isa<AsyncResetType>())
-        resolveResult =
-            refResolveBuilder.create<AsAsyncResetPrimOp>(refResolve);
+    ImplicitLocOpBuilder sinkBuilder(wireModule.getLoc(), wireModule);
```

Comment:
I think this could be `auto sinkBuilder = ImplicitLocOpBuilder::atBlockEnd(targetOp->getBlock())` and moved after the `targetOp` to save a line.